File: app/src/main/java/com/jet/learnq/model/Dictionary.java
package com.jet.learnq.model;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import com.jet.learnq.controller.SQLiteDatabaseController;
import com.jet.learnq.dto.WordDTO;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class Dictionary implements Serializable {
    List<WordModel> dictionary;
    List<WordModel> reversedDictionary;

    List<WordDTO> wordDTOS;
    List<WordDTO> reversedWordDTOS;

    SQLiteDatabaseController database;
    List<LanguageModel> languageModels;
    List<List<WordModel>> pairs;
    SharedPreferences preferences;
    boolean changed;

    public Dictionary(SQLiteDatabaseController database, Context context) {
        dictionary = new ArrayList<>();
        reversedDictionary = new ArrayList<>();
        wordDTOS = new ArrayList<>();
        reversedWordDTOS = new ArrayList<>();
        this.database = database;
        languageModels = database.getAllLanguages();
        preferences = context.getSharedPreferences("Properties", Context.MODE_PRIVATE);
        reloadDictionary();
    }

    public boolean editSet(List<WordModel> changed) {
        return true;
    }

    public void deleteFromDatabase(WordDTO toDeleteDTO,
                                   String languageOn,
                                   String languageTo) {

        database.deleteFromTable(toDeleteDTO, languageOn, languageTo);
        reloadDictionary();
    }

    public List<WordDTO> getDictionary() {
        if (changed) reloadDictionary();
        return wordDTOS;
    }

    public List<WordDTO> getReversedDictionary() {
        if (changed) reloadDictionary();
        return reversedWordDTOS;
    }

    public List<LanguageModel> getLanguageModels() {
        return languageModels;
    }

    public void addANewPair(String word, String translation) {
        database.addOnePair(
                new WordModel(
                        1,
                        database.getLanguageId(
                                preferences.getString(
                                        "default_language_on", "Error")), word, new ArrayList<>()),
                new WordModel(
                        1,
                        database.getLanguageId(
                                preferences.getString(
                                        "default_language_to", "Error")), translation, new ArrayList<>()));
        changed = true;
    }

    private void reloadDictionary() {
        changed = false;
        wordDTOS.clear();
        reversedWordDTOS.clear();
        pairs = database.getAllPairs(
                preferences.getString("default_language_on", "English"),
                preferences.getString("default_language_to", "English"));
        dictionary = pairs.get(0);
        Log.println(Log.INFO, "update", "dic size " + dictionary.size());
        reversedDictionary = pairs.get(1);
        Log.println(Log.INFO, "update", "r dic size " + reversedDictionary.size());

        for (WordModel wm : dictionary) {
            wordDTOS.add(new WordDTO(wm.getTranslations(), wm.getName()));
        }
        for (WordModel wm : reversedDictionary) {
            reversedWordDTOS.add(new WordDTO(wm.getTranslations(), wm.getName()));
        }
    }
}

File: app/src/main/kotlin/com/jet/learnq/Dictionary.kt
package com.jet.learnq

import android.content.Context
import android.content.SharedPreferences
import com.jet.learnq.controller.SQLiteDatabaseController
import com.jet.learnq.dto.WordDTO
import com.jet.learnq.model.LanguageModel
import com.jet.learnq.model.WordModel
import java.util.*

class Dictionary(private val database: SQLiteDatabaseController, context: Context) {
    private var dictionary: MutableList<WordModel> = ArrayList()
    private var reversedDictionary: MutableList<WordModel> = ArrayList()
    private var wordDTOS: MutableList<WordDTO> = ArrayList()
    private var reversedWordDTOS: MutableList<WordDTO> = ArrayList()
    private var languageModels: MutableList<LanguageModel> = ArrayList()
    private var pairs: MutableList<MutableList<WordModel>> = ArrayList()
    private var preferences: SharedPreferences
    private var changed = true

    init {
        preferences = context.getSharedPreferences("Properties", Context.MODE_PRIVATE)
        languageModels = database.allLanguages // TODO сделать его обратно синглтоном
    }

    fun fillPairs(items: List<WordDTO>, stringItem: MutableList<String>) {
        stringItem.clear()
        for (wd: WordDTO in items) {
            var contains = false
            for (s: String in stringItem) {
                contains = s.lowercase(Locale.getDefault()).startsWith(wd.name.lowercase(Locale.getDefault()))
                if (contains) {
                    break
                }
            }
            if (!contains) {
                val stringBuilder = StringBuilder(
                    wd.name + " - " +
                        wd.translations[0].name
                )
                if (wd.translations.size > 1) {
                    for (i in 1 until wd.translations.size) {
                        stringBuilder.append(", ").append(wd.translations[i].name)
                    }
                }
                stringItem.add(stringBuilder.toString())
            }
        }
    }

    fun deleteFromDatabase(
        toDeleteDTO: WordDTO,
        languageOn: String,
        languageTo: String
    ) {
        database.deleteFromTable(toDeleteDTO, languageOn, languageTo)
        reloadDictionary()
    }

    fun getDictionary(): List<WordDTO> {
        if (changed) reloadDictionary()
        return wordDTOS
    }

    fun getReversedDictionary(): List<WordDTO> {
        if (changed) reloadDictionary()
        return reversedWordDTOS
    }

    fun getLanguageModels(): List<LanguageModel> {
        return languageModels
    }

    fun addANewPair(word: String, translation: String) {
        database.addOnePair(
            WordModel(
                1,
                database.getLanguageId(
                    preferences.getString(
                        "default_language_on",
                        "Error"
                    )
                ),
                word,
                ArrayList()
            ),
            WordModel(
                1,
                database.getLanguageId(
                    preferences.getString(
                        "default_language_to",
                        "Error"
                    )
                ),
                translation,
                ArrayList()
            )
        )
    }

    private fun reloadDictionary() {
        changed = false
        wordDTOS.clear()
        reversedWordDTOS.clear()
        val languageOn = preferences.getString("default_language_on", "English")
        val languageTo = preferences.getString("default_language_to", "English")
        pairs = database.getAllPairs(
            languageOn,
            languageTo
        )
        dictionary = pairs[0]
        reversedDictionary = pairs[1]
        for (wm in dictionary) {
            wordDTOS.add(WordDTO(wm.translations, wm.name, languageOn))
        }
        for (wm in reversedDictionary) {
            reversedWordDTOS.add(WordDTO(wm.translations, wm.name, languageTo))
        }
    }
}
